moveit_ros/trajectory_cache/src/trajectory_cache.cpp

```diff
@@ -0,0 +1,1340 @@
+
+    coll.insert(trajectory, insert_metadata);
+    return true;
+  }
+
+  RCLCPP_INFO(node_->get_logger(),
+              "Skipping plan insert: New trajectory execution_time (%es) "
+              "is worse than best trajectory's execution_time (%es)",
+              execution_time_s, best_execution_time);
+  return false;
+}
+
+// MOTION PLAN TRAJECTORY CACHING: QUERY CONSTRUCTION
+bool TrajectoryCache::extract_and_append_trajectory_start_to_query(
+    Query& query, const moveit::planning_interface::MoveGroupInterface& move_group,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request, double match_tolerance)
+{
+  match_tolerance += exact_match_precision_;
+
+  // Make ignored members explicit
+  if (!plan_request.start_state.multi_dof_joint_state.joint_names.empty())
+  {
+    RCLCPP_WARN(node_->get_logger(), "Ignoring start_state.multi_dof_joint_states: Not supported.");
+  }
+  if (!plan_request.start_state.attached_collision_objects.empty())
+  {
+    RCLCPP_WARN(node_->get_logger(), "Ignoring start_state.attached_collision_objects: Not supported.");
+  }
+
+  // auto original = *query;  // Copy not supported.
+
+  query.append("group_name", plan_request.group_name);
+
+  // Workspace params
+  // Match anything within our specified workspace limits.
+  query.append("workspace_parameters.header.frame_id", plan_request.workspace_parameters.header.frame_id);
+  query.appendGTE("workspace_parameters.min_corner.x", plan_request.workspace_parameters.min_corner.x);
+  query.appendGTE("workspace_parameters.min_corner.y", plan_request.workspace_parameters.min_corner.y);
+  query.appendGTE("workspace_parameters.min_corner.z", plan_request.workspace_parameters.min_corner.z);
+  query.appendLTE("workspace_parameters.max_corner.x", plan_request.workspace_parameters.max_corner.x);
+  query.appendLTE("workspace_parameters.max_corner.y", plan_request.workspace_parameters.max_corner.y);
+  query.appendLTE("workspace_parameters.max_corner.z", plan_request.workspace_parameters.max_corner.z);
+
+  // Joint state
+  //   Only accounts for joint_state position. Ignores velocity and effort.
+  if (plan_request.start_state.is_diff)
+  {
+    // If plan request states that the start_state is_diff, then we need to get
+    // the current state from the move_group.
+
+    // NOTE: methyldragon -
+    //   I think if is_diff is on, the joint states will not be populated in all
+    //   of our motion plan requests? If this isn't the case we might need to
+    //   apply the joint states as offsets as well.
+    //
+    // TODO: Since MoveIt also potentially does another getCurrentState() call
+    //   when planning, there is a chance that the current state in the cache
+    //   differs from the state used in MoveIt's plan.
+    //
+    //   When upstreaming this class to MoveIt, this issue should go away once
+    //   the class is used within the move group's Plan call.
+    moveit::core::RobotStatePtr current_state = move_group.getCurrentState();
```

Comment:
The cache methods are meant to mimic the planning methods of movegroup (in terms of their method signature)

I.e. instead of calling
```
move_group.computeCartesianPath(...);
move_group.plan(...);

// Instead:
cache.fetch_best_matching_trajectory(move_group, ...);
cache.fetch_best_matching_cartesian_trajectory(move_group, ...);
```

Additionally, I think the management of current states should be an implementation detail that the caller shouldn't need to think about, at this level of the interface 🤔 (this is especially because current state is only needed under certain circumstances, depending on the properties of the request (`is_diff == true`), which the user normally is not expected to manage.)

What do you think?